### PR TITLE
Implementation of Active CMAES

### PR DESCRIPTION
**Pull request by SuvarshaChennareddy**
This pull request is an initial implementation of Active CMAES, which is part of my Google Summer of Code (GSoC) project. Tests have not been added yet, but they will be included soon.

### CODE REVIEW

**Comment by SuvarshaChennareddy** [comment]
It's just the CMAES tests that are failing.

File: include/ensmallen_bits/cmaes/active_cmaes_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+#ifndef ENSMALLEN_ACTIVE_CMAES_CMAES_IMPL_HPP
+#define ENSMALLEN_ACTIVE_CMAES_CMAES_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "active_cmaes.hpp"
+
+#include <ensmallen_bits/function.hpp>
+
+#ifndef NOT_EMPTY_TRANSFORMATION
+#define NOT_EMPTY_TRANSFORMATION
```
Comment (conradsnicta) [inline include/ensmallen_bits/cmaes/active_cmaes_impl.hpp:24]: not sure what this define is for

File: include/ensmallen_bits/cmaes/cmaes_impl.hpp [hunk 2]
```diff
@@ -20,31 +20,82 @@
 
 #include <ensmallen_bits/function.hpp>
 
+#ifndef NOT_EMPTY_TRANSFORMATION
+#define NOT_EMPTY_TRANSFORMATION
```
Comment (conradsnicta) [inline include/ensmallen_bits/cmaes/cmaes_impl.hpp:24]: not sure what this define is for

File: HISTORY.md [hunk 3]
```diff
@@ -1,5 +1,8 @@
 ### ensmallen ?.??.?: "???"
 ###### ????-??-??
+ * Fix the current CMAES inconsistencies
+   ([#351](https://github.com/mlpack/ensmallen/pull/351)).
```
Comment (conradsnicta) [inline HISTORY.md:4]: this doesn't look right.  the present PR is 367, and has the description of "Implementation of Active CMAES".  why mention 351 here?

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp [hunk 4]
```diff
@@ -423,9 +423,9 @@ inline bool NSGA2::Dominates(
 //! Assign crowding distance to the population.
 template <typename MatType>
 inline void NSGA2::CrowdingDistanceAssignment(
-	const std::vector<size_t>& front,
-	std::vector<arma::Col<typename MatType::elem_type>>& calculatedObjectives,
-	std::vector<typename MatType::elem_type>& crowdingDistance)
+    const std::vector<size_t>& front,
+    std::vector<arma::Col<typename MatType::elem_type>>& calculatedObjectives,
+    std::vector<typename MatType::elem_type>& crowdingDistance)
```
Comment (conradsnicta) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:428]: these are just changes in whitespace. what's the reason for this?  if the changes are not related to Active CMAES, please omit them

**Review by conradsnicta — COMMENTED** [review]
minor comments

File: include/ensmallen_bits/cmaes/active_cmaes_impl.hpp (hunk 1 above)
Comment (SuvarshaChennareddy) [reply include/ensmallen_bits/cmaes/active_cmaes_impl.hpp:24]: Thank you for pointing this out! Now that I look at it, I don't think this define is required. The `ACTIVE_CMAES_CMAES_IMPL_HPP` define should be enough. I'll make changes.

File: include/ensmallen_bits/cmaes/cmaes_impl.hpp (hunk 2 above)
Comment (SuvarshaChennareddy) [reply include/ensmallen_bits/cmaes/cmaes_impl.hpp:24]: I'll be sure to update this too.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp (hunk 4 above)
Comment (rcurtin) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:428]: It fixes the style to only indent twice on a wrapped line---personally I think it's just fine.

File: HISTORY.md (hunk 3 above)
Comment (SuvarshaChennareddy) [reply HISTORY.md:4]: Sorry about that! The branch of this PR (367) is actually a branch of PR 351's branch 
😅. I was trying to find a cleaner way to open this PR, but couldn't really find one (that being said, is there a way?). 
I know I'll have to deal with the merge conflicts if PR 351 gets merged.

Comment (SuvarshaChennareddy) [reply HISTORY.md:4]: I'll update this soon too.

Comment (conradsnicta) [reply HISTORY.md:4]: suggest to keep PR 351 and this PR as separate (ie. not inter-dependent).  if this PR is merged first, PR 351 may then need to be updated (and vice-versa).

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp (hunk 4 above)
Comment (conradsnicta) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:428]: The whitespace changes are not a problem by themselves.  The issue is that this PR seems to be touching stuff that's not directly related to it.  I'm worried that there may be other irrelevant files that are also changed, possibly introducing regressions.

File: HISTORY.md (hunk 3 above)
Comment (SuvarshaChennareddy) [reply HISTORY.md:4]: I understand, I believe (between the two if any) PR 351 will get merged first. So I would have to update this PR soon (resolving merge and conflicts and such).

File: include/ensmallen_bits/cmaes/active_cmaes_impl.hpp (hunk 1 above)
Comment (SuvarshaChennareddy) [reply include/ensmallen_bits/cmaes/active_cmaes_impl.hpp:24]: I'm looking at this once more and I realize that if the define is removed and the user uses both Vanilla and Active CMAES, there will be double inclusion (which is why I added the include guard in the first place). The `ACTIVE_CMAES_CMAES_IMPL_HPP` define is indeed not enough.

Comment (rcurtin) [reply include/ensmallen_bits/cmaes/active_cmaes_impl.hpp:24]: It sounds like in that case that either `NotEmptyTransfornation` should be moved to its own file and included by both versions, or if the functionality differs here then perhaps it just needs a new name. I think I have understood the problem, let me know if not.

Comment (SuvarshaChennareddy) [reply include/ensmallen_bits/cmaes/active_cmaes_impl.hpp:24]: Alright, I'll move `NotEmptyTransfornation` to its own file.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp (hunk 4 above)
Comment (SuvarshaChennareddy) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:428]: These changes were pushed with the last PR [351](https://github.com/mlpack/ensmallen/pull/367).

Comment (zoq) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:428]: Thanks

**Comment by zoq** [comment]
@mlpack-jenkins test this please

**Review by zoq — APPROVED** [review]
No more comments from my side, thanks for the contribution.

**Comment by SuvarshaChennareddy** [comment]
Thank you, @zoq!
There is one thing I would like to address. Wouldn't removing the deprecation go against [this](https://github.com/mlpack/ensmallen/pull/351#discussion_r1064832640). I also think we should make changes (fixing style and changing element types) the to Vanilla CMAES for consistency.

**Comment by SuvarshaChennareddy** [comment]
> If we use `BoundaryBoxConstraint` as the default transformation policy wouldn't that be the same as using the other constructor?

Yes, but the default is `EmptyTransformation`.

**Comment by zoq** [comment]
Do you see any problem with using `BoundaryBoxConstraint` as the default for the constructor?

**Comment by zoq** [comment]
We could even provide another constructor for thee empty transformation.

**Comment by zoq** [comment]
Looking at it again, if we keep the constructor, it's backward compatible, since I can still use the "old" constructor.

**Comment by SuvarshaChennareddy** [comment]
> Looking at it again, if we keep the constructor, it's backward compatible, since I can still use the "old" constructor.

Yes, that's why it was marked as deprecated instead of being removed altogether. We could make `BoundaryBoxConstraint` the default constructor; however, CMA-ES itself isn't bounded. So, I think it would make more sense to keep `EmptyTransformation` as the default. Since it would be the default, keeping it deprecated would also make sense, wouldn't it? These are just my thoughts.

**Comment by zoq** [comment]
I think both interface are fine, the old interface is closer to what we provide for the other optimizers, so I like to keep it, the new one offers the ability to support more transformations so I like to keep it as well.

**Comment by SuvarshaChennareddy** [comment]
Alright. We should probably modify Vanilla CMA-ES as well then (the element types should be changed as well for consistency).

**Comment by zoq** [comment]
Makes sense, I'll resolve the failing test and merge this PR.

**Comment by zoq** [comment]
Thanks for the great contribution.
